src/rcc/clocking.rs

```diff
@@ -0,0 +1,292 @@
+//! The `clocking` module contains representations of the various objects in the STM32L4x6
+//! clock tree (see Reference Manual Figs 15 and 16) useful for wiring them up.
+//!
+//! There are two main concepts: sources (enum variants) and clocks (structs). For example, the
+//! SYSCLK clock can be driven by any one of four sources: HSI, MSI, HSI, and PLLCLK. This
+//! knowledge is encoded in the SysClkSources enum.
+//!
+//! Each enum variant contains information about the clock it represents. Some clocks are
+//! configurable, and thus have fields, and some are not. For example, the LSI
+//! (LowSpeedInternalRC) clock is always 32 kHz, but the MSI (MediumSpeedInternalRC) clock can
+//! be configured and thus has a frequency component.
+//!
+//! To use them, compose them and feed them to, e.g., sysclk.
+//!
+//! ```rust
+//! let mut rcc = RCC.constrain();
+//! let msi_clk = clocking::MediumSpeedInternalRC::new(8_000_000, false);
+//! let sys_clk_src = clocking::SysClkSource::MSI(msi_clk);
+//! let cfgr = rcc.cfgr.sysclk(sys_clk_src);
+//! ```
+//!
+//! The PLL is a bit more complex because it _is_ a source (`PLLClkOutput`) and also _requires_
+//! a source (`PLLClkSource`), but you compose the types similarly.
+
+use super::rcc;
+
+pub trait InputClock {
+    fn freq(&self) -> u32;
+}
+
+/// High-speed internal 16 MHz RC
+#[derive(Clone, Copy)]
+pub struct HighSpeedInternal16RC {
+    pub always_on: bool,
+    pub auto_start: bool,
+}
+
+impl HighSpeedInternal16RC {
+    /// Applies the selection options to the configuration registers and turns the clock on
+    pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+        rcc.cr.modify(|_, w| {
+            w.hsion()
+                .set_bit()
+                .hsikeron()
+                .bit(self.always_on)
+                .hsiasfs()
+                .bit(self.auto_start)
+        });
+        while rcc.cr.read().hsirdy().bit_is_clear() {}
+        (16_000_000, 0b01)
+    }
+}
+
+/// Medium-speed internal 100 kHz - 48 MHz RC
+#[derive(Clone, Copy)]
+pub struct MediumSpeedInternalRC {
+    freq: u32,
+    auto_cal: bool,
+}
+
+impl MediumSpeedInternalRC {
+    pub fn new(freq: u32, auto_cal: bool) -> Self {
+        MediumSpeedInternalRC { freq, auto_cal }
```

Comment:
Should we maybe validate frequency right here?

Reply:
I made a TODO note here - it's a small change but I'll have to get to it later